File: lab1/C++/src/main.cpp
#include <iostream>
#include <vector>

using namespace std;

void merge(int arr[], int aux[], int left, int right) {
    int middle = left + (right - left) / 2;
    int i = left, j = middle + 1, current = left;
    for (int k = left; k <= right; k++) aux[k] = arr[k];
    
    while (i <= middle && j <= right) {
        if (aux[i] <= aux[j]) {
            arr[current++] = aux[i++];
        }
        else {
            arr[current++] = aux[j++];
        }
    }
    while (i <= middle) arr[current++] = aux[i++];
    while (j <= right) arr[current++] = aux[j++];
}

void mergeSort(int arr[], int aux[], int left, int right) {
    if (left >= right) return;
    int middle = left + (right - left) / 2;
    mergeSort(arr, aux, left, middle);
    mergeSort(arr, aux, middle + 1, right);
    merge(arr, aux, left, right);
}

int main() {
    int n;
    int * arr = new int[n];
    int * aux = new int[n];

    cin >> n;
    for (int i = 0; i != n; i++) {
        cin >> arr[i];
    }

    mergeSort(arr, aux, 0, n - 1);
    for (int i = 0; i != n; i++) {
        cout << arr[i] << " ";
    }
    cout << endl;
    delete [] arr;
    delete [] aux;
    return 0;
}

File: lab1/Python/src/main.py
def merge(leftarr, rightarr):
    left, right = 0, 0
    ret = []
    while left < len(leftarr) and right < len(rightarr):
        if leftarr[left] < rightarr[right]:
            ret.append(leftarr[left])
            left = left + 1
        else:
            ret.append(rightarr[right])
            right = right + 1
    ret += leftarr[left:]
    ret += rightarr[right:]
    return ret


def mergeSort(arr, left, right):
    if right - left <= 1:
        return arr
    middle = left + (right - left) // 2
    arr = mergeSort(arr, left, middle)
    arr = mergeSort(arr, middle + 1, right)
    return merge(arr[left:middle+1], arr[middle+1:right+1])


if __name__ == '__main__':
    filename = 'in.in'
    file = open(filename)
    arr = []
    for line in file.readlines():
        arr.append(eval(line))
    arr = arr[1::]
    arr = mergeSort(arr, 0, len(arr) - 1)
    flag = True
    for i in range(1, len(arr)):
        if arr[i] < arr[i - 1]:
            print(f"wrong ans at {i}, where arr[{i-1}] = {arr[i - 1]}, arr[{i}] = {arr[i]}")
            flag = False
    if flag:
        print("OK!")
